File: python/day3.py
import re

with open("../../inputs/d3", "r") as f:
    input = f.read()
    input.replace("\n", "")


def part1(input):
    reg = r"mul\(([0-9]+),([0-9]+)\)"
    acc = 0
    for mul in re.finditer(reg, input):
        acc += int(mul.group(1)) * int(mul.group(2))
    return acc


def part2(input):
    reg = r"(?<=(^|do\(\()(.(?!don't\(\)))*)mul\((\d+),(\d+)\)"
    acc = 0
    for mul in re.finditer(reg, input):
        print(mul)
        # acc += int(mul.group(1)) * int(mul.group(2))


ans1 = part1(input)
ans2 = part2(input)
print(ans1, ans2)

assert ans1 == 163931492

assert ans2 == 76911921

File: elixir/day3.exs
defmodule Day3 do
  def parse(content) do
    String.trim(content)
  end

  def part1(input) do
    Regex.scan(~r/mul\((\d+),(\d+)\)/, input)
      |> Enum.map(fn [_, x, y] -> {String.to_integer(x), String.to_integer(y)} end)
      |> Enum.reduce(0, fn {x, y}, acc -> acc + x * y end)
  end

  def part2(input) do
    Regex.scan(~r/(?>mul\((\d+),(\d+)\))|(?>do\(\))||(?>don't\(\))/, input)
      |>Enum.filter(&(&1 != [""]))
      |>Enum.reduce({true, 0}, fn token, {continue?, total} ->
        case {continue?, token} do
          {_, ["do()"]} -> {true, total}
          {_, ["don't()"]} -> {false, total}
          {true, token} -> {true, total + multiply(token)}
          {false, _} -> {false, total}
        end
      end)
      |>elem(1)
  end

  def multiply([_ | op]) do
    String.to_integer(Enum.at(op, 0)) * String.to_integer(Enum.at(op, 1))

  end
end

data = "../input/day3.txt"
  |> File.read!()
  |> Day3.parse()

IO.puts "Part1: #{Day3.part1(data)}"
IO.puts "Part2: #{Day3.part2(data)}"
